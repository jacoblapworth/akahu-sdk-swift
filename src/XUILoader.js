import React from 'react';
import cn from 'classnames';
import CSSClasses from 'xui-css-classes';

const sizeMap = {
	standard: CSSClasses.Loader.BASE,
	large: CSSClasses.Loader.LARGE,
	small: CSSClasses.Loader.SMALL
};

const propTypes = {
	className: React.PropTypes.string,
	qaHook: React.PropTypes.string,
	label: React.PropTypes.string,
	defaultLayout: React.PropTypes.bool,
	size: React.PropTypes.oneOf(Object.keys(sizeMap))
};

const defaultProps = {
	defaultLayout: true,
	size: 'standard'
};

const XUILoader = (props) => {
	const sizeClass = sizeMap[props.size];
	const className = cn(CSSClasses.Loader.BASE, sizeClass, props.className, (props.defaultLayout ? CSSClasses.Loader.LAYOUT : ''));

	return (
		<div data-automationid={props.qaHook} className={className} role="progressbar" aria-label={props.label} aria-busy="true">
			<div className={CSSClasses.Loader.DOT} />
			<div className={CSSClasses.Loader.DOT} />
			<div className={CSSClasses.Loader.DOT} />
		</div>
	)
};

XUILoader.propTypes = propTypes;
XUILoader.defaultProps = defaultProps;
export default XUILoader;
