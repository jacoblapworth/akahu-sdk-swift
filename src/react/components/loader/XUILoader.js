import PropTypes from 'prop-types';
import React from 'react';
import cn from 'classnames';
import '../helpers/xuiGlobalChecks';
import { sizeClassNames, baseClass } from './private/constants';

const dot = <div className={`${baseClass}--dot`} />;

const XUILoader = props => {
	const className = cn(
		baseClass,
		sizeClassNames[props.size],
		props.className,
		(props.defaultLayout && !props.retainLayout) && `${baseClass}-layout`,
		props.isInverted && `${baseClass}-inverted`,
		props.retainLayout && `${baseClass}-retain-layout`,
	);

	return (
		<div
			data-automationid={props.qaHook}
			className={className}
			role="progressbar"
			aria-label={props.ariaLabel}
		>
			{dot}{dot}{dot}
		</div>
	);
};

XUILoader.propTypes = {
	/** Add additional classes to the loader wrapping div */
	className: PropTypes.string,

	/** Adds data-automationid attribute with qaHook contents to the loader wrapping div */
	qaHook: PropTypes.string,

	/** adds aria-label to the loader wrapping div */
	ariaLabel: PropTypes.string,

	/** Defaults to `true`. Sets the default layout class on the loader wrapping div */
	defaultLayout: PropTypes.bool,

	/** Sets the size of the loader to be, medium (default), small, or xsmall */
	size: PropTypes.oneOf(Object.keys(sizeClassNames)),

	/** Sets the loader to the inverted colour scheme */
	isInverted: PropTypes.bool,

	/** Adds the retain layout class, used in combination with buttons. Applying this prop
	 * will cause `defaultLayout` prop to be ignored. */
	retainLayout: PropTypes.bool,
};

XUILoader.defaultProps = {
	defaultLayout: true,
	size: 'medium',
	ariaLabel: 'Loading',
};

export default XUILoader;
