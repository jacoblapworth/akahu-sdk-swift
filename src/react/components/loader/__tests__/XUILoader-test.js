import React from 'react';
import { mount } from 'enzyme';
import XUILoader from '../XUILoader';

describe('XUILoader', function () {
	it('should render with an aria label describing its purpose', function () {
		const testString = 'Something is loading, please wait';
		const wrapper = mount(<XUILoader label={testString} />);
		expect(wrapper.getDOMNode().getAttribute('aria-label')).toEqual(testString);
	});

	it('should add extra classes when defined', function () {
		const testClass = 'test-class';
		const wrapper = mount(<XUILoader className={testClass} label="Something is loading, please wait" />);
		expect(wrapper.getDOMNode().classList.contains(testClass)).toBeTruthy();
	});

	it('should add the layout class by default', function () {
		const testClass = 'test-class';
		const wrapper = mount(<XUILoader className={testClass} label="Something is loading, please wait" />);
		expect(wrapper.getDOMNode().classList.contains('xui-loader-layout')).toBeTruthy();
	});

	it('should not add the layout class if `defaultLayout` is set to `false`', function () {
		const wrapper = mount(<XUILoader defaultLayout={false} label="Something is loading, please wait" />);
		expect(wrapper.getDOMNode().classList.contains('xui-loader-layout')).toBeFalsy();
	});

	it('should add appropriate size classes', function () {
		const wrapper = mount(
			<div>
				<XUILoader label="Something is loading, please wait" />
				<XUILoader size="small" label="Something is loading, please wait" />
				<XUILoader size="large" label="Something is loading, please wait" />
			</div>
		);

		const loaders = wrapper.find('.xui-loader-layout');
		expect(loaders.at(0).hasClass('xui-loader')).toBeTruthy();
		expect(loaders.at(1).hasClass('xui-loader-small')).toBeTruthy();
		expect(loaders.at(2).hasClass('xui-loader-large')).toBeTruthy();
	});

	it('should add the inverted class if `isInverted` is set to `true`', () => {
		const wrapper = mount(
			<XUILoader
				isInverted={true}
				label="Something is loading, please wait" />
		);

		expect(wrapper.getDOMNode().classList.contains('xui-loader-inverted')).toBeTruthy();
	});

	it('should add the retain layout class if `retainLayout` is set to `true`', function () {
		const wrapper = mount(
			<XUILoader
				retainLayout={true}
				label="Something is loading, please wait" />
		);
		expect(wrapper.getDOMNode().classList.contains('xui-loader-retain-layout')).toBeTruthy();
	});
});
