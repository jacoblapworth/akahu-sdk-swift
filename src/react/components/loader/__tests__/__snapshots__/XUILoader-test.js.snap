// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XUILoader should add appropriate size classes 1`] = `
<div>
  <div
    aria-label="Something is loading, please wait"
    className="xui-loader xui-loader-standard xui-loader-layout"
    role="progressbar"
  >
    <div
      className="xui-loader--dot"
    />
    <div
      className="xui-loader--dot"
    />
    <div
      className="xui-loader--dot"
    />
  </div>
  <div
    aria-label="Something is loading, please wait"
    className="xui-loader xui-loader-small xui-loader-layout"
    role="progressbar"
  >
    <div
      className="xui-loader--dot"
    />
    <div
      className="xui-loader--dot"
    />
    <div
      className="xui-loader--dot"
    />
  </div>
  <div
    aria-label="Something is loading, please wait"
    className="xui-loader xui-loader-xsmall xui-loader-layout"
    role="progressbar"
  >
    <div
      className="xui-loader--dot"
    />
    <div
      className="xui-loader--dot"
    />
    <div
      className="xui-loader--dot"
    />
  </div>
</div>
`;
